base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1838,75 +1865,104 @@ fn handle_possible_reorg<T: BlockchainBackend>(
-        new_block.header.height,
-        new_block_hash.to_hex(),
-        db_height,
-        new_tips.len()
+        "Validate and add {} chain block(s) from block {}. Rewound blocks: [{}]",
+        chain.len(),
+        fork_hash.to_hex(),
+        removed_blocks
+            .iter()
+            .map(|b| b.height().to_string())
+            .collect::<Vec<_>>()
+            .join(", ")
     );
 
-    if new_tips.is_empty() {
-        debug!(
-            target: LOG_TARGET,
-            "No reorg required, could not construct complete chain using block #{} ({}).",
-            new_block.header.height,
-            new_block_hash.to_hex()
-        );
-        return Ok(BlockAddResult::OrphanBlock);
+    for block in chain {
+        let mut txn = DbTransaction::new();
+        let block_hash = *block.hash();
+        txn.delete_orphan(block_hash);
+        let chain_metadata = backend.fetch_chain_metadata()?;
+        if let Err(e) = block_validator.validate_body_for_valid_orphan(backend, block, &chain_metadata) {
+            warn!(
+                target: LOG_TARGET,
+                "Orphan block {} ({}) failed validation during chain reorg: {:?}",
+                block.header().height,
+                block_hash.to_hex(),
+                e
+            );
+            remove_orphan(backend, block_hash)?;
+
+            info!(target: LOG_TARGET, "Restoring previous chain after failed reorg.");
+            restore_reorged_chain(backend, fork_hash, removed_blocks)?;
+            return Err(e.into());
+        }
+
+        insert_best_block(&mut txn, block.clone())?;
+        // Failed to store the block - this should typically never happen unless there is a bug in the validator
+        // (e.g. does not catch a double spend). In any case, we still need to restore the chain to a
+        // good state before returning.
+        if let Err(e) = backend.write(txn) {
+            warn!(
+                target: LOG_TARGET,
+                "Failed to commit reorg chain: {:?}. Restoring last chain.", e
+            );
+
+            restore_reorged_chain(backend, fork_hash, removed_blocks)?;
+            return Err(e);
+        }
     }
 
+    Ok(removed_blocks)
+}
+
+fn swap_to_highest_pow_chain<T: BlockchainBackend>(
+    db: &mut T,
+    config: &BlockchainDatabaseConfig,
+    block_validator: &dyn PostOrphanBodyValidation<T>,
+    chain_strength_comparer: &dyn ChainStrengthComparer,
+) -> Result<BlockAddResult, ChainStorageError> {
+    let metadata = db.fetch_chain_metadata()?;
+    // lets clear out all remaining headers that dont have a matching block
+    // rewind to height will first delete the headers, then try delete from blocks, if we call this to the current
+    // height it will only trim the extra headers with no blocks
+    rewind_to_height(db, metadata.height_of_longest_chain())?;
+    let all_orphan_tips = db.fetch_all_orphan_chain_tips()?;
+    if all_orphan_tips.is_empty() {
```

Comment:
Maybe this check should be before we rewind

Reply:
I specifically did not do this as I use the rewind code here to remove the "extra" headers. 
I want to "main chain" to be "clean" after this run. 

The intended case is this. 

Header sync downloads 10 headers. 
Block sync fails after 5. 
That extra 5 headers will stay on top of the main chain till we remove them again. I want to remove them at this point regardless of whether the new tip is higher than our original one.